src/FMBot.Bot/Services/IndexService.cs

```diff
@@ -193,26 +216,82 @@ public async Task<GuildUser> GetOrAddUserToGuild(Persistence.Domain.Models.Guild
+                               "WHERE guild_id = @guildId AND user_id = @userId ";
+
+            DefaultTypeMap.MatchNamesWithUnderscores = true;
+            await using var connection = new NpgsqlConnection(this._botSettings.Database.ConnectionString);
+            await connection.OpenAsync();
+
+            var guild = (await connection.QueryAsync<Persistence.Domain.Models.Guild>("SELECT * FROM guilds WHERE guild_id = @guildId", new
+            {
+                guildId
+            }))
+            .ToList().FirstOrDefault();
+
+            var dto = new Models.IndexedUserUpdateDto
+            {
+                UserName = discordName,
+                GuildId = guildId,
+                UserId = userId,
+            };
+
+            if (guild.WhoKnowsWhitelistRoleId.HasValue)
+            {
+                dto.WhoKnowsWhitelisted = discordGuildUser.RoleIds.Contains(guild.WhoKnowsWhitelistRoleId.Value);
+            }
+
+            await connection.ExecuteAsync(sql, dto);
+        }
+
+        public async Task UpdateDiscordUser(IGuildUser discordGuildUser)
         {
             var discordName = discordGuildUser.Nickname ?? discordGuildUser.Username;
 
             await using var db = this._contextFactory.CreateDbContext();
 
             const string sql = "UPDATE guild_users " +
-                               "SET user_name =  @userName " +
+                               "SET user_name =  @userName, " +
+                               "who_knows_whitelisted =  @whoKnowsWhitelisted " +
                                "WHERE guild_id = @guildId AND user_id = @userId ";
 
             DefaultTypeMap.MatchNamesWithUnderscores = true;
             await using var connection = new NpgsqlConnection(this._botSettings.Database.ConnectionString);
             await connection.OpenAsync();
 
-            await connection.ExecuteAsync(sql, new
+            var guild = (await connection.QueryAsync<Persistence.Domain.Models.Guild>("SELECT * FROM guilds WHERE discord_guild_id = @discordGuildId", new
+            {
+                DiscordGuildId = discordGuildUser.GuildId,
+            }))
+            .ToList().FirstOrDefault();
+
+            var user = (await connection.QueryAsync<GuildUser>("SELECT * FROM users WHERE discord_user_id = @discordUserId", new
+            {
+                DiscordUserId = discordGuildUser.Id,
+            }))
+            .ToList().FirstOrDefault();
+
+            var dto = new Models.IndexedUserUpdateDto
+            {
+                UserName = discordName,
+                GuildId = guild.GuildId,
+                UserId = user.UserId,
+            };
+
+            if (guild.WhoKnowsWhitelistRoleId.HasValue)
             {
-                userName = discordName,
-                guildId = guildId,
-                userId = userId
-            });
+                dto.WhoKnowsWhitelisted = discordGuildUser.RoleIds.Contains(guild.WhoKnowsWhitelistRoleId.Value);
```

Comment:
does this work if the user has no roles at all?

Reply:
Works on my instance